Add MKdocs for Friendica doc

nupplaphil opened the pull request:
Because of our recent discussion, I selected a popular, small Open Source framework (https://www.mkdocs.org/) to generate a  more pretty, better searchable and easy translatable output.

With the current config, which is far from ready, the output would currently look like that: https://doc-test.philipp.info/

So we could setup a https://doc.friendi.ca domain and with woodpecker, it would be easy to automatically update the docs :-)
We could even create versioned docs (develop vs. stable)

The only "disadvantage" would be that we have to rewrite the `/help` sub-directory of the Friendica nodes. But tbh I don't find them THAT structured and a little bit clumpy to use. So in the end, it would be possible to automatically generate the static output into a "help" sub-folder as well.

ToDos:
- [x] Restructure current docs
- [x] Add missing German translation skeleton files
- [x] Add CI auto-generation
- [x] Use generated docs for Friendica node `/help` 
- [x] Use generated docs for (versioned) deployment onto https://doc.friendi.ca ( pinging for help with the new domain:  @annando and @utzer :)
- [x] adapt `DBStructure` methods because of new paths (thx @annando )

nupplaphil committed:
Add MKdocs for Friendica doc

MrPetovan commented:
None of the links of https://doc-test.philipp.info/ seem to work at the moment except the tab links in the top menu. 😅

In particular, the edit link doesn't work, see how it looks: https://github.com/friendica/friendica/edit/develop/doc/en/Bugs-and-Issues.md

Quix0r commented:
I guess it far from being finished. https://doc-test.philipp.info/developer/Tests/?h=test actually works here (I used the search). So I guess we have to be more patient.

nupplaphil commented:
Yes sorry, it's really far from ready. That's why it's at WIP state :-)

The navbar on the left should work and the translation of these sites.

But this is a big work to do and I want to know first if you agree with it, so @annando , @tobiasd as well :-)

Concerning Translation:
Every new language can be added as a new root dir beside the existing one and as far as it has the same directory structure and name, the switch at the top will work with one additional config line :)

Theming (fonts, colors,...) Are changeable too, even multitheming is supported :)

foss- commented:
Thanks for taking action on this based on the recent discussion. 2 Cents: personal preference for `ReadTheDocs` theme.

nupplaphil commented:
> Thanks for taking action on this based on the recent discussion. 2 Cents: personal preference for `ReadTheDocs` theme.

Thank you! Do you see any advantages for `ReadTheDocs` over `Mkdocs` with `material`? Because material is, as far as I can see, better integrated into `Mkdocs`, but there's a community edition of `ReadTheDocs`, but it seems there is no on-premise solution - just a basic fork for themen `Mkdocs` like `ReadTheDocs` (see https://github.com/cjsheets/mkdocs-rtd-dropdown)

So if we would choose this theme, from my point of view, there has to be some real advantages *gg* ..

[Edit] OK there are more features at ReadTheDocs --> https://docs.readthedocs.io/en/stable/features.html but tbh I'd like to stick at `Mkdocs`... But please challenge my point of view @foss- if you see problems/disatvantages

nupplaphil commented:
Updates :-)

Changes:
- Navigation is almost complete
- German syntax/grammar fixes
- Added possibility for tags
- Added fixed anchor links
- New Examples:
  - https://doc-test.philipp.info/developer/addon-storage-backend/#icanconfigurestorage
  - https://doc-test.philipp.info/de/admin/install/

annando commented:
Please have in mind that the database structure files are created automatically via this function: https://github.com/friendica/friendica/blob/e6ed8f6315d770a244406629d88669f63a5689d0/src/Database/DBStructure.php#L163

nupplaphil commented:
Yes, it's possible to have language agnostic files, so I guess it shouldn't be a problem :)

annando commented:
My comment was meant as a hint that you please change the paths in that function as well :grimacing:

nupplaphil commented:
Yes got it 😊

nupplaphil commented:
@annando the index, db-structure and navigation is now automatically created, have a look at:

https://doc-test.philipp.info/spec/database/

my adapted `DBStructure` code is still uggly I must confess, but it's currently working :)

annando commented:
I just had a thought: Possibly we can automatically generate the API documentation from the router definition?

nupplaphil commented:
I like your thought!! :D

foss- commented:
> But please challenge my point of view @foss- if you see problems/disatvantages

If overhead is increased using ReadTheDocs theme, lets go with the theme that causes less work.

nupplaphil committed:
Add CreateDoc class

Quix0r commented on src/Console/CreateDoc.php:
```diff
@@ -0,0 +1,165 @@
+	 *
+	 * @return string added doxygen header
+	 */
+	private function addDocumentation($line)
```
Can you please add type-hints, both `string` here.

nupplaphil replied:
not relevant anymore

Quix0r commented on src/Console/CreateDoc.php:
```diff
@@ -0,0 +1,165 @@
+{
+	protected $helpOptions = ['h', 'help', '?'];
+
+	protected function getHelp()
```
Can you please add documentation here?

nupplaphil replied:
This is a standard procedure for executing tasks, it's an override of the  base class .. I added a `/** {@inheritDoc} */` to mark it .)

Quix0r commented on src/Console/CreateDoc.php:
```diff
@@ -0,0 +1,165 @@
+		return $help;
+	}
+
+	protected function doExecute()
```
Can you please add documentation here?

nupplaphil replied:
This is a standard procedure for executing tasks, it's an override of the  base class .. I added a `/** {@inheritDoc} */` to mark it .)

nupplaphil committed:
- add help directory & adapt english content
- add woodpecker CI
- replace all help-links

Quix0r reviewed (changes requested):
Please repeat this on all other occurrences, too! Or should I make a separate PR for this? @MrPetovan

nupplaphil commented:
> Please repeat this on all other occurrences, too! Or should I make a separate PR for this? @MrPetovan

Please adapt it at a separate PR .. this one is confusing/big enough :D

Quix0r commented:
@nupplaphil I would recommend a `git rebase upstream/develop` (in case you call this repository "upstream").

nupplaphil commented:
automatically built at https://docs.friendi.ca/develop .. still WIP ..

and the next pr should be a correction of content, which is out of date in some cases..

nupplaphil commented:
... done :-)

Addons, Hooks etc. are part of followups

The messages.po and license check are buggy .. another thing for followups ;-)

nupplaphil committed:
add computed help

MrPetovan commented:
Thank you for your work, what's the workflow to add to the documentation going forward?

nupplaphil commented:
good point .. I'll add a new doc at `doc` for developers :))

nupplaphil committed:
add changes because of rebase

nupplaphil committed:
Adapt doc structure

nupplaphil committed:
restructures and add default friendica.css for later adaptions

nupplaphil committed:
try pushing ..

nupplaphil committed:
remove help again

nupplaphil committed:
try it ..

nupplaphil committed:
fix url

nupplaphil committed:
etc hosts ..

nupplaphil committed:
url .. once again

nupplaphil committed:
url .. once again

nupplaphil committed:
use variable ..

MrPetovan commented:
Please rebase on the latest develop given the forecasted conflicts.

MrPetovan commented on mkdocs.yml:
```diff
@@ -0,0 +1,264 @@
+#### DON'T EDIT THIS FILE AT ROOT, use /view/templates/mkdocs.yaml.tpl Template ####
```
This file has been moved since
```suggestion
#### DON'T EDIT THIS FILE AT ROOT, use /view/templates/doc/mkdocs.yaml.tpl Template ####
```

MrPetovan commented on view/templates/doc/mkdocs.yml.tpl:
```diff
@@ -0,0 +1,191 @@
+#### DON'T EDIT THIS FILE AT ROOT, use /view/templates/mkdocs.yaml.tpl Template ####
```
```suggestion
#### DON'T EDIT THIS FILE AT ROOT, use /view/templates/doc/mkdocs.yaml.tpl Template ####
```

MrPetovan commented:
@nupplaphil Do you mind me asking why you closed this PR?

foss- commented:
Personally had high hopes for this to resolve the clusterjoy the friendica wiki currently is. What is the alternative route forward?

nupplaphil commented:
It was by accident, thx for mentioning it :-)

tobiasd commented:
In case this PR shall be part of the 2023.09 (10) release, please redirect it towards the 2023.09-rc branch.

MrPetovan commented:
And now it's full of conflicts, do you need help rebasing this, @nupplaphil ?

foss- commented:
Polite ping. This would be much needed. Popped back into my mind, since a third party developer linked https://docs.friendi.ca/develop/spec/api/mastodon/#implemented-endpoints

This is already so much more usable than the current wiki. It looks great, the way information is shown is very accessible.

I'd be glad if it was made usable and replaced the current wiki sooner than later.

MrPetovan commented:
I'll try to resolve the conflicts this week to move this PR forward. Recent questions about the outdated wikis made this effort to create a standalone documentation website based on the most up-to-date documentation in the MD files more important than ever.

Art4 commented:
@nupplaphil May I ask if I could take over this PR and fix the git conflicts?

MrPetovan commented:
I started working on it, one of the main problems is that we changed the terminology for "groups" and "forums" into "circles" and "groups", which affected some file paths. Additionally, @nupplaphil added some language changes to the pages themselves on top of changing the formatting.

Since the branch is made of a single large commit, I'm not sure how to commit my existing work to enable you to take over.

@nupplaphil Did you use an automated tool to update the formatting of the help pages or did you do it by hand?

Art4 commented:
Maybe we have to start over again, but split the PR into multiple parts:

1. Create only the `mkdocs.yml` file and the fundamental folder structure with 1-2 docs files
2. Migrate existing docs file. This way we can better talk about the new folder structure. Maybe we can improve it?
3. Migrate the contents of the Wiki (if appropriate).

MrPetovan commented:
If you plan on starting from scratch, the little I already did is meaningless and you can take over!

Art4 commented:
@nupplaphil If thats OK for you I will start with a fresh PR once the RC branch for 2025.02 is created.

nupplaphil commented:
Yes it's ok, in February, i'm still really busy, maybe in march. But if you need help, don't hesitate. I can add the pipeline for creating it automatically at docs.friendi.ca

marcusxss commented:
A few months later than expected, but the 2025.07 branch has been created! :partying_face:

annando commented:
Shouldn't we set this to draft, since it isn't ready to be merged in this stage?
